src/eos-updater-avahi.c

```diff
@@ -0,0 +1,659 @@
+
+  return value != NULL;
+}
+
+EosAvahiDiscoverer *
+eos_avahi_discoverer_new (GMainContext *context,
+                          EosAvahiDiscovererCallback callback,
+                          gpointer user_data,
+                          GDestroyNotify notify,
+                          GError **error)
+{
+  g_autoptr(EosAvahiDiscoverer) discoverer = NULL;
+
+  g_return_val_if_fail (callback != NULL, NULL);
+  g_return_val_if_fail (error == NULL || *error == NULL, NULL);
+
+  discoverer = g_object_new (EOS_TYPE_AVAHI_DISCOVERER, NULL);
+  discoverer->callback = callback;
+  discoverer->user_data = user_data;
+  discoverer->notify = notify;
+  if (context == NULL)
+    discoverer->context = g_main_context_ref_thread_default ();
+  else
+    discoverer->context = g_main_context_ref (context);
+  discoverer->discovered_services = g_hash_table_new_full (g_str_hash,
+                                                           g_str_equal,
+                                                           g_free,
+                                                           NULL);
+  discoverer->found_services = object_array_new ();
+
+  if (use_avahi_emulator ())
+    {
+      if (!setup_emulated_avahi_discoverer (discoverer, error))
+        return NULL;
+    }
+  else if (!setup_real_avahi_discoverer (discoverer, context, error))
+    return NULL;
+
+  return g_steal_pointer (&discoverer);
+}
+
+static gchar *
+txt_records_to_string (const gchar **txt_records)
+{
+  g_autoptr(GString) str = NULL;
+  gsize idx;
+
+  str = g_string_new ("");
+
+  for (idx = 0; txt_records[idx] != NULL; idx++)
+    {
+      g_autofree gchar *record_escaped = g_markup_escape_text (txt_records[idx], -1);
+      g_string_append_printf (str, "    <txt-record>%s</txt-record>\n",
+                              record_escaped);
+    }
+
+  return g_string_free (g_steal_pointer (&str), FALSE);
+}
+
+static GBytes *
+generate_from_avahi_service_template (const gchar *type,
+                                      guint16 port,
+                                      const gchar *txt_version,
+                                      const gchar **txt_records)
+{
+  g_autofree gchar *service_group = NULL;
+  gsize service_group_len;  /* bytes, not including trailing nul */
+  g_autofree gchar *txt_records_str = txt_records_to_string (txt_records);
+  g_autofree gchar *type_escaped = g_markup_escape_text (type, -1);
+  g_autofree gchar *txt_version_escaped = g_markup_escape_text (txt_version, -1);
+
+  service_group = g_strdup_printf (
+      "<service-group>\n"
+      "  <name replace-wildcards=\"yes\">EOS update service on %%h</name>\n"
+      "  <service>\n"
+      "    <type>%s</type>\n"
+      "    <port>%" G_GUINT16_FORMAT "</port>\n"
+      "    <txt-record>eos_txt_version=%s</txt-record>\n"
+      "%s"
+      "  </service>\n"
+      "</service-group>\n",
+      type_escaped, port, txt_version_escaped, txt_records_str);
```

Comment:
It is correct to advertise the service if `eos-update-server` is not running, since `eos-update-server` is socket-activated.

See my comments above for my thoughts so far about this. I haven’t made any changes to fix this yet, and will think about it some more tomorrow.